cat: change Column to a struct and related improvements

RaduBerinde opened the pull request:
#### cat: remove unneeded ColTypeStr API

Release note: None

#### cat: change Column to a struct

The opt catalog currently  exposes ` ColumnDescriptors` as `cat.Column`s.  The
intent here was to avoid allocating more objects, but with the addition of
system and virtual columns this approach has outlived its usefulness. This
design is limiting because we can't add more fields, like the `Kind`.

This change switches `optTable` and ` optVirtualTable` to storing all columns in
a single slice. Because we no longer have any real implementations of
`cat.Column`, we change it to a struct. The struct has private members and the
same getter interface. This removes the need to have multiple similar
implementations and avoids vtable dispatch for all these getters.

Release note: None

#### cat: remove Table.ColumnKind

Now that we have `Column.Kind()`, we no longer need the awkward `ColumnKind` on
the table.

Release note: None

#### cat: add Column.Ordinal()

This change adds the column ordinal to `cat.Column`, and removes the ordinal in
`cat.IndexColumn` (which can now be accessed as `Ordinal()` thanks to the
embedded `*Column`).

Release note: None

RaduBerinde committed:
cat: change Column to a struct

The opt catalog currently exposes `ColumnDescriptors` as `cat.Column`s.  The
intent here was to avoid allocating more objects, but with the addition of
system and virtual columns this approach has outlived its usefulness. This
design is limiting because we can't add more fields, like the `Kind`.

This change switches `optTable` and `optVirtualTable` to storing all columns in
a single slice. Because we no longer have any real implementations of
`cat.Column`, we change it to a struct. The struct has private members and the
same getter interface. This removes the need to have multiple similar
implementations and avoids vtable dispatch for all these getters.

Release note: None

RaduBerinde committed:
cat: remove Table.ColumnKind

Now that we have `Column.Kind()`, we no longer need the awkward `ColumnKind` on
the table.

Release note: None

RaduBerinde committed:
cat: add Column.Ordinal()

This change adds the column ordinal to `cat.Column`, and removes the ordinal in
`cat.IndexColumn` (which can now be accessed as `Ordinal()` thanks to the
embedded `*Column`).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52846)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, this cleaned up nice. Thanks for doing the work.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-:-MEkJ56D98t2-W2_HjBj:b-5s64ru)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/sql/opt_catalog.go, line 717 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkH-f7DefohgTiRLso:-MEkH-f7DefohgTiRLsp:b-1txkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt_catalog.go#L717)):*
> ```Go
> 			// TODO(radu, mjibson): figure out what the type should be here. Geo is
> 			// Int, but JSON isn't anything decodable (including Bytes). The disk
> 			// fetecher will need to be taught about inverted indexes and dump the
> ```

NIT: fetecher => fetcher
___
*[pkg/sql/opt/cat/column.go, line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFQ7NCKMv5SBzaTMm:-MEkFQ7NCKMv5SBzaTMn:b-kthk1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L106)):*
> ```Go
> //
> // TODO(andyk): Switch calling code to use DatumType.
> func (c *Column) ColTypeWidth() int {
> ```

Do you mind using `DatumType().Width()` here instead? That would save us on size of column. This `ColTypeWidth` and `ColTypePrecision` fields predate the conversion to `types.T`, and I've been meaning to remove them for a long while.
___
*[pkg/sql/opt/cat/column.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFxSp9WnCm4o20y1v:-MEkFxSp9WnCm4o20y1w:b-c004af) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L240)):*
> ```Go
> 			panic(errors.AssertionFailedf("column type should never be a nested array"))
> 		}
> 		t = t.ArrayContents()
> ```

This is really surprising that we'd use the type of `ArrayContents`. Can you add a comment explaining why? I'd think this will cause bugs, and am now scratching my head about how this would ever have worked.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/cat/column.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFxSp9WnCm4o20y1v:-MEkOFML-e-T8275jY_C:blfut9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L240)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This is really surprising that we'd use the type of `ArrayContents`. Can you add a comment explaining why? I'd think this will cause bugs, and am now scratching my head about how this would ever have worked.
</blockquote></details>

I have no idea.. I did what the existing code does (we're only using it for the precision and width).


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52846)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/sql/opt/cat/column.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFxSp9WnCm4o20y1v:-MEkRTxG2tjszFupHX0r:b-hh4q22) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L240)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I have no idea.. I did what the existing code does (we're only using it for the precision and width).
</blockquote></details>

I tried removing the special handling in a separate branch and all logictests pass. I will post a separate PR with that (and then rebase this one).


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52846)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @RaduBerinde)

___
*[pkg/sql/opt/cat/column.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFxSp9WnCm4o20y1v:-MEkRORL94sWmpJv7akR:b-wd4mnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L240)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I tried removing the special handling in a separate branch and all logictests pass. I will post a separate PR with that (and then rebase this one).
</blockquote></details>

Oh I see. For some reason I was reading the code as if `c.datumType = t.ArrayContents()`. This does complicate things if we get rid of the precision and width methods.  But it looks like there's only 1 usage of each method, so better to put logic like this there, rather than assuming every caller wants precision/width of the array contents.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52846)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @ajwerner, @andy-kimball, and @RaduBerinde)

___
*[pkg/sql/opt/cat/column.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-MEkFxSp9WnCm4o20y1v:-MEkWQWI0hYC0p-fKijF:b-k2l31f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6306ad183f55dc4548ec0165f424d28be110d484/pkg/sql/opt/cat/column.go#L240)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Oh I see. For some reason I was reading the code as if `c.datumType = t.ArrayContents()`. This does complicate things if we get rid of the precision and width methods.  But it looks like there's only 1 usage of each method, so better to put logic like this there, rather than assuming every caller wants precision/width of the array contents.
</blockquote></details>

Ah, ok. Since there's no behavior change in just moving the logic, I added it to this PR as a separate commit.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
cat: remove ColTypeWidth and ColTypePrecision

Removing these cat.Column methods; it is simpler to derive them from the
`DatumType()` in the one place they are used.

Release note: None

RaduBerinde committed:
cat: remove unneeded ColTypeStr API

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 20 files at r6, 7 of 15 files at r7, 21 of 21 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52846#-:-MEy1MWr50f5Vl4Du7Fc:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @andy-kimball, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

nvanbenschoten commented:
I think this PR skewed with `master`, but only under race compilation, which is why it wasn't caught. Sending a patch to fix.
